packages/db/src/indexes/basic-index.ts

```diff
@@ -0,0 +1,439 @@
+    // which doesn't matter for Set
+    return this.rangeQuery(options)
+  }
+
+  /**
+   * Returns the next n items in sorted order
+   */
+  take(n: number, from?: any, filterFn?: (key: TKey) => boolean): Array<TKey> {
+    const result: Array<TKey> = []
+
+    let startIdx = 0
+    if (from !== undefined) {
+      const normalizedFrom = normalizeValue(from)
+      startIdx = this.findInsertionIndex(normalizedFrom)
+      // Skip past the 'from' value (exclusive)
+      while (
+        startIdx < this.sortedValues.length &&
+        this.compareFn(this.sortedValues[startIdx], normalizedFrom) <= 0
+      ) {
+        startIdx++
+      }
+    }
+
+    for (
+      let i = startIdx;
+      i < this.sortedValues.length && result.length < n;
+      i++
+    ) {
+      const keys = this.valueMap.get(this.sortedValues[i])
+      if (keys) {
+        for (const key of keys) {
+          if (result.length >= n) break
+          if (!filterFn || filterFn(key)) {
+            result.push(key)
+          }
+        }
+      }
+    }
+
+    return result
+  }
+
+  /**
+   * Returns the next n items in reverse sorted order
+   */
+  takeReversed(
+    n: number,
+    from?: any,
+    filterFn?: (key: TKey) => boolean
+  ): Array<TKey> {
+    const result: Array<TKey> = []
+
+    let startIdx = this.sortedValues.length - 1
+    if (from !== undefined) {
+      const normalizedFrom = normalizeValue(from)
+      startIdx = this.findInsertionIndex(normalizedFrom) - 1
+      // Skip past the 'from' value (exclusive)
+      while (
+        startIdx >= 0 &&
+        this.compareFn(this.sortedValues[startIdx], normalizedFrom) >= 0
+      ) {
+        startIdx--
+      }
+    }
+
+    for (let i = startIdx; i >= 0 && result.length < n; i--) {
+      const keys = this.valueMap.get(this.sortedValues[i])
+      if (keys) {
+        for (const key of keys) {
+          if (result.length >= n) break
+          if (!filterFn || filterFn(key)) {
+            result.push(key)
+          }
+        }
+      }
+    }
+
+    return result
+  }
+
+  /**
+   * Performs an IN array lookup - O(k) where k is values.length
+   */
+  inArrayLookup(values: Array<any>): Set<TKey> {
+    const result = new Set<TKey>()
+
+    for (const value of values) {
+      const normalizedValue = normalizeValue(value)
+      const keys = this.valueMap.get(normalizedValue)
+      if (keys) {
+        keys.forEach((key) => result.add(key))
+      }
+    }
+
+    return result
+  }
+
+  // Getter methods for testing/compatibility
+  get indexedKeysSet(): Set<TKey> {
+    return this.indexedKeys
+  }
+
+  get orderedEntriesArray(): Array<[any, Set<TKey>]> {
+    return this.sortedValues.map((value) => [
+      value,
+      this.valueMap.get(value) ?? new Set(),
+    ])
+  }
+
+  get orderedEntriesArrayReversed(): Array<[any, Set<TKey>]> {
+    return [...this.sortedValues]
+      .reverse()
+      .map((value) => [value, this.valueMap.get(value) ?? new Set()])
```

Comment:
This code iterates 3 times through the sorted values. We can improve it by a single pass with a for loop or with `reduce`.